### PR TITLE
Use jemalloc in rocksdb

### PR DESCRIPTION
**Pull request by akhi3030**
We’re using jemalloc in Rust code so it makes sense to do the same in
RocksDB.

Fixes: https://github.com/near/nearcore/issues/6290

### CODE REVIEW

**Comment by mina86** [comment]
This should be enabled only if we enable jemalloc feature on neard

**Comment by akhi3030** [comment]
> This should be enabled only if we enable jemalloc feature on neard

Isn't it already using it?  https://github.com/near/nearcore/blob/master/neard/Cargo.toml#L28

**Comment by nikurt** [comment]
Do we understand the performance impact of this change?

**Comment by akhi3030** [comment]
> Do we understand the performance impact of this change?

Nope, not yet.  Do we have some standard set of tests that we can use?

**Comment by nikurt** [comment]
> Do we have some standard set of tests that we can use?

That's what I did last time:
https://near.zulipchat.com/#narrow/stream/297873-pagoda.2Fnode/topic/Performance.20impact.20of.20logging/near/283804107

**Comment by mina86** [comment]
> > This should be enabled only if we enable jemalloc feature on neard
> 
> Isn't it already using it? https://github.com/near/nearcore/blob/master/neard/Cargo.toml#L28

I mean this feature https://github.com/near/nearcore/blob/master/neard/Cargo.toml#L47 should enable jemalloc for RocksDB.

File: core/store/Cargo.toml [hunk 1]
```diff
@@ -14,7 +14,13 @@ bytesize = { version = "1.1", features = ["serde"] }
 derive_more = "0.99.3"
 elastic-array = "0.11"
 enum-map = "2.1.0"
-rocksdb = { version = "0.18.0", default-features = false, features = ["snappy", "lz4", "zstd", "zlib"] }
+rocksdb = { version = "0.18.0", default-features = false, features = [
```
Comment (matklad) [inline core/store/Cargo.toml:17]: I think we have rocks in three different Cargo.tomls, each needs to be updated

**Comment by matklad** [comment]
> Do we have some standard set of tests that we can use?

@akhi3030 another approach came up recently here: https://github.com/near/nearcore/pull/7494#issuecomment-1231385870

File: core/store/Cargo.toml (hunk 1 above)
Comment (mina86) [reply core/store/Cargo.toml:17]: I’m waiting for Cargo 1.64 to fix that.

Comment (akhi3030) [reply core/store/Cargo.toml:18]: Updated all cargo.tomls that mention rocksdb.

**Review by matklad — APPROVED** [review]
LGTM, and I am inclined to merge this without blocking on up-front benchmark:

* the status quo is that we don't know which allocator rocksdb is useing, 'cause it links the one dynamically from the computer neard is running on
* any significant perf degradation should be visible in our continous estimation
* we won't immediately notice change in memory usage, but I think it unlikely to worsen, and we can always revert if we observe something iffy on the betanet.

**Comment by akhi3030** [comment]
Makes sense.  I'll merge.

**Comment by mina86** [comment]
I still think this isn’t correct.  We have a cargo feature in neard/Cargo.toml to enable jemalloc.  IMO RocksDB jemalloc should be conditionally enabled if that feature is enabled.

**Comment by akhi3030** [comment]
Can you point me to the relevant bit of code please?

**Comment by mina86** [comment]
https://github.com/near/nearcore/blob/master/neard/Cargo.toml#L47

PS. An alternative is to get rid of the feature and always use jemalloc I suppose.  Have we actually ever performed any benchmarks to see if jemalloc is faster for us?

**Comment by matklad** [comment]
Ah, good catch, thanks mina86! Yeah, I think we should just enable jemalloc by default: it's not even the question of "faster", without jemalloc we don't know which allocator we are using.

I don't remeber us performing benchmark for nearcore, but, eg, for rust-analyzer and rust-lang, we did observe jemalloc being measurably faster, and, in ra, we did spend some cycles chasing down issue which boiled down to "the user's allocator is not good".

**Pull request merged**